README.md
#teamwall

| Branch | Status |
--------|--------
| Master | [[![Build Status](https://travis-ci.org/teamwall/teamwall.svg?branch=master)](https://travis-ci.org/teamwall/teamwall)](https://travis-ci.org/teamwall/teamwall) |
| Develop | [[![Build Status](https://travis-ci.org/teamwall/teamwall.svg?branch=develop)](https://travis-ci.org/teamwall/teamwall)](https://travis-ci.org/teamwall/teamwall)|

An open source software used to build a wall of picture for your team

## Contents

- [Prerequisites](#prerequisites)
- [Running](#running)
- [Server configuration](#server-configuration)
    - [config subcommand](#config-subcommand)
    - [CLI arguments](#cli-arguments)
- [API](#api)
    - [Register a new user](#register-a-new-user)
        - [Path & Verb](#path-verb)
        - [Parameters](#parameters)
        - [Status](#status)
    - [Login](#login)
        - [Path & Verb](#path-verb)
        - [Parameters](#parameters)
        - [Response](#response)
        - [Status](#status)
    - [Get current user](#get-current-user)
        - [Path & Verb](#path-verb)
        - [Parameters](#parameters)
        - [Response](#response)
        - [Status](#status)
    - [Update user settings](#update-user-settings)
        - [Path & Verb](#path-verb)
        - [Parameters](#parameters)
        - [Status](#status)
    - [Get team members](#get-team-members)
        - [Path & Verb](#path-verb)
        - [Parameters](#parameters)
        - [Response](#response)
        - [Status](#status)
    - [Upload a new photo](#upload-a-new-photo)
        - [Path & Verb](#path-verb)
        - [Parameters](#parameters)
        - [Response](#response)
        - [Status](#status)
    - [Get last photo of a user](#get-last-photo-of-a-user)
        - [Path & Verb](#path-verb)
        - [Parameters](#parameters)
        - [Response](#response)
        - [Status](#status)
- [License](#license)

<!-- end toc -->

## Prerequisites

You will need [Leiningen][1] 2.0.0 or above installed.

You will also need [Mongo][2] database running.

Other dependencies can be installed running:

    $ bower install

[1]: https://github.com/technomancy/leiningen
[2]: http://www.mongodb.org/

## Running

The first time you want to run the application,
you first need to build the client part. To do so, run:

    $ lein setup

To start a web server for the application, run:

    $ lein run

The web server should be reachable at `localhost:3000`.

By default, _Teamwall_ is served using HTTP but it is highly recommended to serve it using HTTPS.
This can be easily done using [Nginx](http://nginx.org/en/) as detailled [here](https://github.com/teamwall/teamwall/wiki/HTTPS).

## Server configuration

There is two kind of settings which can be provided to _Teamwall_.
The first kind contains the settings that can be stored into the databse.
The second kind contains the databases settings.

### config subcommand

The `config` subcommand can be used to set the port used to serve the application.

    $ lein run config --port PORT

The help text can be seen using the command:

    $ lein run config --help

### CLI arguments

This settings concern mainly the database access.
The list of available options can be found running: 

    $ lein run -- --help

## API

### Register a new user

#### Path & Verb

To register a new mate, a **POST** request needs to be done against the path:

    /register

#### Parameters

The payload must be JSON.

| Params | Type | Description|
--------|--------|--------
| email | payload | The new user email. It should be unique |
| username | payload | The new user user name |
| password | payload | The new user password |

cURL example:

    $ curl -H "Content-Type: application/json" -d '{"email":"user1@mycompany.com", "username": "John Doe", "password": "password"}' localhost:3000/register

#### Status

| Cause | Status |
--------|--------
| Success | 200 OK |
| Email already used | 403 Forbidden |

### Login

First step is to login to retrieve an API token which will be used for every other request.

#### Path & Verb

To login, a **GET** request needs to be done against the path:

    /login

#### Parameters

| Params | Type | Description|
--------|--------|--------
| email | url param | the email address used to register|
| password | url param | the password used to register|

cURL example:

    $ curl "localhost:3000/login?email=user1@mycompany.com&password=password"

#### Response

A JSON string representing a literal object with three keys:

* token: a string representing the unique API token
* user: a map representing the current user
* ttl: the token time-to-live

Example:

~~~json
{
	"token" : "JSzvjh_Qq0zGjLu7pL-9tvGrl84DwgMNT4vZ_F4IxC",
	"ttl"   : "3600000"
}
~~~

Note: _ttl_ is in milliseconds, `3600000` is one hour.

#### Status

| Cause | Status |
--------|--------
| Success | 200 OK |
| Error | 403 Forbidden |

### Get current user

The route is used to provide information about the current logged user.
It is mainly used when you log in back using the cookie stored token.

#### Path & Verb

To get all the team members, a **GET** request needs to be done against the path:

    /current-user

#### Parameters

| Params | Type | Description|
--------|--------|--------
| token | url param | the API token received during the login |

cURL example:

    $ curl "localhost:3000/current-user?token=JSzvjh_Qq0zGjLu7pL-9tvGrl84DwgMNT4vZ_F4IxC"


#### Response

A JSON string representing a literal object.
This object has two keys:

* username
* email

Example:

~~~json
{
  "username" : "User 1",
  "email" : "user1@mycompany.com"
}
~~~

#### Status

| Cause | Status |
--------|--------
| Success | 200 OK |
| Field _token_ missing or invalid | 403 Forbidden |

### Update user settings

#### Path & Verb

To update the current user's settings, a **POST** request needs to be done against the path:

    /settings

#### Parameters

The payload must be JSON.

| Params | Type | Description|
--------|--------|--------
| token | payload | the API token received during the login |
| settings | payload | a json map of the user settings |

cURL example:

    $ curl -H "Content-Type: application/json" -d '{"sleep-time": 60, "token":"JSzvjh_Qq0zGjLu7pL-9tvGrl84DwgMNT4vZ_F4IxC"}' localhost:3000/settings

#### Status

| Cause | Status |
--------|--------
| Success | 200 OK |
| Field _token_ missing or invalid | 403 Forbidden |

### Get team members

The route is used to provided information about the members of the team the logged user belongs to.
The team is bound by a shared email domain.

#### Path & Verb

To get all the team members, a **GET** request needs to be done against the path:

    /team-members

#### Parameters

| Params | Type | Description|
--------|--------|--------
| token | url param | the API token received during the login |

cURL example:

    $ curl "localhost:3000/team-members?token=JSzvjh_Qq0zGjLu7pL-9tvGrl84DwgMNT4vZ_F4IxC"


#### Response

A JSON string representing an array of literal objects.
Each object has two keys:

* username
* email

Example:

~~~json
[ {
  "username" : "User 1",
  "email" : "user1@mycompany.com"
}, {
  "username" : "User 2",
  "email" : "user2@mycompany.com"
} ]
~~~

#### Status

| Cause | Status |
--------|--------
| Success | 200 OK |
| Field _token_ missing or invalid | 403 Forbidden |

### Upload a new photo

Route used to upload a new photo for the user linked to the provided token.

#### Path & Verb

To upload a new photo, a **POST** request needs to be done against the path:

    /new-photo

#### Parameters

| Params | Type | Description|
--------|--------|---------
| token | url param | the API token received during the login |
| photo | multipart/form-data | the new photo to upload |

cURL example:

    $ curl --form photo=@photo.png "localhost:3000/new-photo?token=JSzvjh_Qq0zGjLu7pL-9tvGrl84DwgMNT4vZ_F4IxC"


#### Response

No response.

#### Status

| Cause | Status |
--------|--------
| Success | 200 OK |
| Field _token_ missing or invalid | 403 Forbidden |
| Field _photo_ missing | 400 Bad Request |

### Get last photo of a user

This route is used to get the last photo of a giver user.
#### Path & Verb

To get the latest photo, a **GET** request needs to be done against the path:

    /<email>/last-photo

#### Parameters

| Params | Type | Description|
--------|--------|---------
| email | url segment | the email address of one of your team member |
| token | url param | the API token received during the login |


cURL example:

    $ curl "http://localhost:3000/user1@mycompany.com/last-photo?token=JSzvjh_Qq0zGjLu7pL-9tvGrl84DwgMNT4vZ_F4IxC"


#### Response

Return an image. The url is made so it can be used as html image source.

#### Status

| Cause | Status |
--------|--------
| Success | 200 OK |
| Email address not in the same team | 400 Bad Request |
| Field _token_ missing or invalid | 403 Forbidden |
| No photo available | 404 Not Found |



## License

Copyright © 2014 Benjamin Van Ryseghem <benjamin [dot] vanryseghem [at] gmail [dot] com>


CHANGES
1.1.0 Adds settings

  * Adds server settings via CLI (TW-35)
  * Adds user settings via a in-app menu (TW-36)
  * Adds logout (TW-48)
  * Adds a link to register page (TW-39)

1.0.2 Fixes jar generation

1.0.1 Fixes the default snapshot time

1.0.0 Initial public release


src/teamwall/handler.clj
(ns teamwall.handler
  (:use [slingshot.slingshot :only [throw+ try+]])
  (:require [cheshire.core :refer :all]
            [clojure.java.io :as io]
            [compojure.core :refer :all]
            [compojure.handler :refer [site]]
            [compojure.route :as route]
            [crypto.random :as random]
            [ring.middleware.multipart-params :refer [wrap-multipart-params]]
            [ring.middleware.params :refer [wrap-params]]
            [teamwall.api :as api]
            [teamwall.db :as db]
            [teamwall.serializer :as serializer]))


;;    /==================\
;;    |                  |
;;    |       VARS       |
;;    |                  |
;;    \==================/


(def ^{:private true} default-ttl
  "Default value for a token Time-To-Live. Default is one hour"
  3600000)
(def ^{:private true} setting-file-name
  "Name of the setting file"
  "settings.tw")
(def ^{:private true} tokens
  "Atom storing all the active tokens"
  (atom {}))

(def ^{:private true} register-token
  "Random token used to check the registration form"
  (atom ""))


;;    /==================\
;;    |                  |
;;    |     PRIVATES     |
;;    |                  |
;;    \==================/


(defn- default-settings
  "Generates a new default setting map"
  []
  {:salt (str (java.util.UUID/randomUUID))})

(defn- check-settings
  "Checks the existence of a setting file"
  []
  (let [file-exists? (.exists (io/as-file setting-file-name))]
    (if file-exists?
      (def ^{:private true} settings
        "Content of the setting file"
        (serializer/read-from-file setting-file-name))
      (do
        (def ^{:private true} settings
          "Content of the setting file"
          (default-settings))
        (serializer/write-in-file settings setting-file-name)))))

(defn- generate-api-token
  "Generates a new API token"
  []
  (random/url-part 32))

(defn- token-valid?
  "Checks that the provided token exists and is still alive"
  [tokens token]
  (if (contains? tokens token)
    (let [data (get tokens token)
          ttl (:ttl data)
          creation-time (:creation-time data)
          now (java.util.Date.)]
      (< (- (.getTime now) (.getTime creation-time)) ttl))
    false))

(defn- login
  "Checks if the user info are correct and generates an API token for the user"
  [email password salt]
  (try+
   (let [user (db/retrieve-user email
                                password
                                salt)
         token (generate-api-token)]
     (swap! tokens assoc token {:user          user
                                :ttl           default-ttl
                                :creation-time (java.util.Date.)})

     {:status  200
      :headers {"Content-Type" "application/json"}
      :body    (generate-string
                {:token token
                 :ttl default-ttl}
                {:pretty true})})
   (catch [:type :teamwall.db/login-failed] {:keys [email valid-password?]}
     {:status 403})
   (catch  Exception e
     (.printStackTrace e)
     {:status 500})))

(defn- secure-routing
  "Encapsulate the check of token validity"
  [token fun]
  (let [valid (token-valid? @tokens token)]
    (if valid
      (try+
       (fun (:user (get @tokens token)))
       (catch [:type ::request-error] {:keys [status]}
         {:status status})
       (catch Exception e
         (.printStackTrace e)
         {:status 500}))
      {:status 403})))

(defn- secure-routing-json
  "Encapsulate the check of token validity and wraps the result in JSON"
  [token fun]
  (secure-routing token
                  (fn [user]
                    (let [result (fun (:user (get @tokens token)))]
                      (if-not (nil? result)
                        {:status  200
                         :headers {"Content-Type" "application/json"}
                         :body    (generate-string
                                   result
                                   {:pretty true})}
                        {:status  200
                         :headers {"Content-Type" "application/json"}})))))


;;    /==================\
;;    |                  |
;;    |      ROUTES      |
;;    |                  |
;;    \==================/


(defroutes app-routes
  "Defines the server routes"
  (GET "/register"
       {}
       (do
         (reset! register-token (generate-api-token))
         (str "<form name=\"input\" action=\"register\" method=\"post\">
              Username: <input type=\"text\" name=\"username\"><br>
              Email: <input type=\"email\" name=\"email\"><br>
              Password: <input type=\"password\" name=\"password\"><br>
              Confirm Password: <input type=\"password\" name=\"cfm-pwd\"><br>
              <input type=\"hidden\" name=\"token\" value="
              @register-token
              ">
              <input type=\"submit\" value=\"Register\">
              </form>")))
  (POST "/register"
        {params :params}
        (let [get-token @register-token]
          (reset! register-token (generate-api-token))
          (when (= get-token (:token params))
            (db/register-user
             (:username params)
             (:password params)
             (:email params)
             (:salt settings))
            "User successfully created")))
  (GET "/login"
       {params :params}
       (login (:email params)
              (:password params)
              (:salt settings)))

  (GET "/team-members"
       {params :params}
       (secure-routing-json (:token params)
                            api/get-team-members))

  (wrap-multipart-params
   (POST "/new-photo"
         {params :params}
         (secure-routing-json (:token params)
                              (fn [user]
                                (if (:photo params)
                                  (api/set-new-photo user
                                                     (:photo params))
                                  (throw+ {:type ::request-error
                                           :status 400}))))))

  (GET "/:email/last-photo"
       {params :params}
       (secure-routing (:token params)
                       (fn [user]
                         (api/last-photo user
                                         (:email params)))))

  (GET "/test"
       {params :params}
       (secure-routing-json (:token params)
                            #({:user %1})))

  (route/resources "/")
  (ANY "/*" [] {:status 403}))

(def app
  "main function used as hook from compojure"
  (site app-routes))


;;    /==================\
;;    |                  |
;;    |       MAIN       |
;;    |                  |
;;    \==================/


(defn- main
  "Initialization of the server"
  []
  (check-settings))

(main)
